### PR TITLE
Limit concurrent Lucene operators

### PR DESCRIPTION
**Pull request by dnhatn**
Currently, when running with segment or document data partitioning, ESQL can enqueue a large number of Lucene operators into the esql_worker threadpool. As a result, queries may fail with rejection due to a high number of outstanding tasks in the threadpool's queue.

This PR restricts the number of Lucene operators per query on each node to the number of threads in the esql_worker threadpool. Additionally, Lucene operators now poll and process Lucene slices from a shared queue. This change also reduces the number of intermediate aggregation states, consequently decreasing the workload of the final aggregation.

### CODE REVIEW

**Comment by dnhatn** [comment]
run elasticsearch-ci/packaging-tests-unix-sample

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSliceQueue.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Shared Lucene slices between Lucene operators.
+ */
+public final class LuceneSliceQueue {
+    private static final int MAX_DOCS_PER_SLICE = 250_000; // copied from IndexSearcher
+    private static final int MAX_SEGMENTS_PER_SLICE = 5; // copied from IndexSearcher
```
Comment (costin) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSliceQueue.java:31]: 👍

**Review by costin — COMMENTED** [review]
+1 on organizing the code better

**Review by nik9000 — APPROVED** [review]
Cool. I didn't look super deeply but it makes sense. Instead of queueing a zillion operations we queue as many as there are cores and have a work-stealing style queue thing.

**Comment by dnhatn** [comment]
@elasticsearchmachine update branch

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSlice.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * Holds a list of multiple partial Lucene segments
+ */
+public final class LuceneSlice {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSlice.java]: Sure, I pushed https://github.com/elastic/elasticsearch/pull/99189/commits/5ac63af32b41c2d454d5b32b21fc89f89423696c

**Comment by dnhatn** [comment]
@costin @nik9000 Thanks for review.

**Pull request merged**